airbyte-commons-worker/src/main/java/io/airbyte/workers/internal/DefaultAirbyteStreamFactory.java

```diff
@@ -34,25 +40,49 @@
 public class DefaultAirbyteStreamFactory implements AirbyteStreamFactory {
 
   private static final Logger LOGGER = LoggerFactory.getLogger(DefaultAirbyteStreamFactory.class);
+  private final double MAX_SIZE_RATIO = 0.8;
 
   private final MdcScope.Builder containerLogMdcBuilder;
   private final AirbyteProtocolPredicate protocolValidator;
   protected final Logger logger;
+  private final long maxMemory;
+  private final Optional<Class<? extends RuntimeException>> exceptionClass;
 
   public DefaultAirbyteStreamFactory() {
     this(MdcScope.DEFAULT_BUILDER);
   }
 
   public DefaultAirbyteStreamFactory(final MdcScope.Builder containerLogMdcBuilder) {
-    this(new AirbyteProtocolPredicate(), LOGGER, containerLogMdcBuilder);
+    this(new AirbyteProtocolPredicate(), LOGGER, containerLogMdcBuilder, Optional.empty());
   }
 
+  /**
+   * Create a default airbyte stream, if a `messageSizeExceptionClass` is not empty, the message size
+   * will be checked and if it more than the available memory * MAX_SIZE_RATIO the sync will be failed
+   * by throwing the exception provided. The exception must have a constructor that accept a string.
+   */
   DefaultAirbyteStreamFactory(final AirbyteProtocolPredicate protocolPredicate,
                               final Logger logger,
-                              final MdcScope.Builder containerLogMdcBuilder) {
+                              final MdcScope.Builder containerLogMdcBuilder,
+                              final Optional<Class<? extends RuntimeException>> messageSizeExceptionClass) {
     protocolValidator = protocolPredicate;
     this.logger = logger;
     this.containerLogMdcBuilder = containerLogMdcBuilder;
+    this.exceptionClass = messageSizeExceptionClass;
+    this.maxMemory = Runtime.getRuntime().maxMemory();
```

Comment:
AFAIK `maxMemory()` returns the total amount of memory reserved for the VM (including the unallocated memory). Judging by the rest of the code, there is an assumption here that 80% of that will be available for local variables - how is this assumption made and validated? 
I might suggest that we use `freeMemory()` instead